scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+        :param name1: axis x name
+        :param name2: axis y name
+        :param align: boolean, match columns in fac1 and fac2 using linear_sum_assignment?
+        """
+
+        corr12 = np.corrcoef(fac1, fac2, False)
+        ind_top = np.arange(0, fac1.shape[1])
+        ind_right = np.arange(0, fac1.shape[1]) + fac1.shape[1]
+        corr12 = corr12[ind_top, :][:, ind_right]
+        corr12[np.isnan(corr12)] = -1
+
+        if align:
+            img = corr12[:, linear_sum_assignment(2 - corr12)[1]]
+        else:
+            img = corr12
+
+        plt.imshow(img)
+
+        plt.title(f'Training initialisation \n {name1} vs {name2}')
+        plt.xlabel(name2)
+        plt.ylabel(name1)
+
+        plt.tight_layout()
+
+        if return_aligned:
+            return linear_sum_assignment(2 - corr12)[1]
+        
+    def plot_posterior_mu_vs_data(self, mu_node_name='mu', data_node='X_data'):
+        r""" Plot expected value of the model (e.g. mean of poisson distribution)
+
+        :param mu_node_name: name of the object slot containing expected value
+        :param data_node: name of the object slot containing data
+        """
+
+        if type(mu_node_name) is str:
+            mu = getattr(self, mu_node_name)
+        else:
+            mu = mu_node_name
+
+        if type(data_node) is str:
+            data_node = getattr(self, data_node)
+
+        plt.hist2d(np.log10(data_node.flatten() + 1), np.log10(mu.flatten() + 1), bins=50,
+                   norm=matplotlib.colors.LogNorm())
+        plt.gca().set_aspect('equal', adjustable='box')
+        plt.xlabel('Data, log10(nUMI)')
+        plt.ylabel('Posterior sample, log10(nUMI)')
+        plt.title('UMI counts (all cell, all genes)')
+        plt.tight_layout()
+        
+    def plot_history(self, iter_start=0, iter_end=-1,
+                     history_key=None, log_y=True, ax=None):
+        r""" Plot training history
+
+        :param iter_start: omit initial iterations from the plot
+        :param iter_end: omit last iterations from the plot
+        """
+
+        if ax is None:
+            ax = plt
+            ax.set_xlabel = plt.xlabel
+            ax.set_ylabel = plt.ylabel
+
+        if history_key is None:
+            history_key = self.history.keys()
+
+        if type(history_key) == str:
+            history_key = [history_key]
+
+        for i in history_key:
+
+            if iter_end == -1:
+                iter_end = np.array(self.history[i]).flatten().shape[0]
+
+            y = np.array(self.history[i]).flatten()[iter_start:iter_end]
+            if log_y:
+                y = np.log10(y)
+            ax.plot(np.arange(iter_start, iter_end), y, label='train')
+            ax.set_xlabel('Training epochs')
+            ax.set_ylabel('Reconstruction accuracy (ELBO loss)')
+            ax.legend()
+            plt.tight_layout()
+            
+            
+    def export2adata(self, adata, slot_name='mod'):
```

Comment:
same here, just return the dict and let users do what they want with it (in tutorial just immediately add it back to anndata)

Reply:
Where would be a good place to intermediately store posterior samples? (nowhere, just give them to users?)